python/activator/middleware_interface.py

```diff
@@ -251,6 +264,33 @@ def _export_calibs(self, export, detector_id, filter):
                                                                         collectionTypes=target_types):
             export.saveCollection(collection)
 
+    def _prep_collections(self):
+        """Pre-register output collections in advance of running the pipeline.
+        """
+        # NOTE: Because we receive a butler on init, we can't use this
+        # prep_butler() because it takes a repo path.
+        # butler = SimplePipelineExecutor.prep_butler(
+        #     self.repo,
+        #     inputs=[self.calibration_collection,
+        #             self.instrument.makeDefaultRawIngestRunName(),
+        #             'refcats'],
+        #     output=self.output_collection)
+        # The below is taken from SimplePipelineExecutor.prep_butler.
+        output_run = f"{self.output_collection}/{self.instrument.makeCollectionTimestamp()}"
+        self.butler.registry.registerCollection(self.instrument.makeDefaultRawIngestRunName(),
+                                                CollectionType.RUN)
+        self.butler.registry.registerCollection(output_run, CollectionType.RUN)
+        self.butler.registry.registerCollection(self.output_collection, CollectionType.CHAINED)
+        collections = [self.calibration_collection,
+                       self.instrument.makeDefaultRawIngestRunName(),
+                       'refcats',
+                       output_run]
```

Comment:
It might be a good unit test for `_prep_butler` [sic] to check if the test input data can all be found in `self.output_collection` with e.g. `butler.datasetExists`. I think such a test wouldn't be affected by the fact that all the files are empty.

Reply:
`output_collection` is a chained collection that includes all its inputs. This is the standard Gen 3 repository structure for doing pipeline runs.